build-system/tasks/visual-diff.js

```diff
@@ -35,10 +100,463 @@ function setPercyBranch() {
+ * set of given webpages.
+ *
+ * @param {!Percy} percy a Percy-Puppeteer controller.
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ * @param {!Array<JsonObject>} webpages an array of JSON objects containing
+ *     details about the pages to snapshot.
+ */
+async function generateSnapshots(percy, page, webpages) {
+  if (argv.master) {
+    await percy.snapshot('Blank page', page, SNAPSHOT_EMPTY_BUILD_OPTIONS);
+  }
+  cleanupAmpConfig();
+
+  const numFlakyTests = webpages.filter(webpage => webpage.flaky).length;
+  if (numFlakyTests > 0) {
+    log('info', 'Skipping', colors.cyan(numFlakyTests), 'flaky tests');
+  }
+  for (const config of CONFIGS) {
+    applyAmpConfig(config);
+    log('verbose',
+        'Generating snapshots using the', colors.cyan(config), 'AMP config');
+    await snapshotWebpages(percy, page, webpages, config);
+  }
+}
+
+/**
+ * Generates Percy snapshots for a set of given webpages.
+ *
+ * @param {!Percy} percy a Percy-Puppeteer controller.
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ * @param {!JsonObject} webpages a JSON objects containing details about the
+ *     pages to snapshot.
+ * @param {*} config Config being used. One of 'canary' or 'prod'.
+ */
+async function snapshotWebpages(percy, page, webpages, config) {
+  webpages = webpages.filter(webpage => (!webpage.flaky &&
+        !webpage.url.startsWith('examples/visual-tests/amp-by-example')));
+  for (const webpage of webpages) {
+    const {url} = webpage;
+    const name = `${webpage.name} (${config})`;
+
+    await enableExperiments(page, webpage['experiments']);
+    log('verbose', 'Navigating to page', colors.yellow(`${BASE_URL}/${url}`));
+    await page.goto(`${BASE_URL}/${url}`);
+
+    await verifyCssElements(page, url, webpage.forbidden_css,
+        webpage.loading_incomplete_css, webpage.loading_complete_css);
+    await percy.snapshot(name, page, SNAPSHOT_OPTIONS);
+    await clearExperiments(page);
+  }
+}
+
+/**
+ * Verifies that all CSS elements are as expected before taking a snapshot.
+ *
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ * @param {string} url URL to be snapshotted.
+ * @param {!Array<string>} forbiddenCss Array of CSS elements that must not be
+ *     found in the page.
+ * @param {!Array<string>} loadingIncompleteCss Array of CSS elements that must
+ *     eventually be removed from the page.
+ * @param {!Array<string>} loadingCompleteCss Array of CSS elements that must
+ *     eventually appear on the page.
+ */
+async function verifyCssElements(page, url, forbiddenCss, loadingIncompleteCss,
+  loadingCompleteCss) {
+  // Wait for loader dot to be hidden.
+  await page.waitForSelector('.i-amphtml-loader-dot', {
+    hidden: true,
+    timeout: CSS_SELECTOR_TIMEOUT_MS,
+  }).catch(() => {
+    log('fatal', colors.cyan(url),
+        `still has the AMP loader dot after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
+  });
```

Comment:
This is good, but as you discovered, loader dots inside iframes could still be visible. Do you think there's a way to look for *all* loader dots to disappear?

Also, in general, this is a place where we should do a more robust job of waiting on the page to fully load. An example of a test page for which waiting made a difference was `amp-gist`.

Feel free to deal with both these items in a separate PR.